Initial port of Python 3.6.0a2

moreati opened the pull request:
The following work
- Invoke the interpreter with cloudabi-run
- Provide python code to execute through argdata
- Import from stdlib modules
- `print()` to stderr

Almost everything else doesn't work yet, or it's untested.

EdSchouten committed:
Check in an initial, but broken Python package.

I've had this package lying around in a private branch for some time,
but hadn't found any time to complete this further. Rebase it on top of
HEAD and clean it up.

Progress on completing this package can be tracked in issue #2.

moreati committed:
Patch handling of thread local storage (TLS) keys

As far as I can tell

 - Python's PyThread assume that TLS keys are always a simple counter,
   the PyThread* functions simply pass around an int.
 - On CloudABI pthread_key_t is actually struct

So as a first stab this patch introduce a new type PyThread_type_key.
This is backward incompatible, so it's just a band-aid - not a proper fix.

moreati committed:
Patch a stub isatty() which always returns false

moreati committed:
Remove all references to stdin & stdout

moreati committed:
#idef references to signal(), SIG_IGN ...

We get signal?
No
Main view... oh

moreati committed:
Remove more references to setlocale() and LC_*

moreati committed:
Remove references to uid_t, gid_t, getuid(), getgid(), ...

moreati committed:
Patch in S_IF* constants

moreati committed:
Declare unsafe string usage in posixmodule.c (strcpy)

moreati committed:
Disable more references to open()

moreati committed:
Disable references to SCHED_* constants

moreati committed:
Disable sole reference to access()

moreati committed:
Disable references to chdir()

moreati committed:
Dsable references to chmod()

moreati committed:
Disable references to fopen()

moreati committed:
Disable references to fork()

moreati committed:
Disable references to getcwd()

moreati committed:
Disable references to kill()

moreati committed:
Disable references to mkdir()

moreati committed:
Disable references to opendir()

moreati committed:
Disable references to pid_t and getpid()

moreati committed:
Disable references to printf(), or use fprintf()

moreati committed:
Disable references to pwd.h

moreati committed:
Disable references to rename()

moreati committed:
Disable references to rmdir()

moreati committed:
Disable references to stat() & lstat()

moreati committed:
Use strtok_r() instead of strtok()

strtok() is not available in cloudlibc, because it it not thread safe.

moreati committed:
Disable references to ttyname()

moreati committed:
Disable references to umask()

moreati committed:
Disable references to unlink()

moreati committed:
Disable references to wait()

moreati committed:
Resolve conflict with other patches

moreati committed:
Python: Don't allow pgen to regenerate grammar graminit.*

The Python tarball ships a Grammar file defining Python syntax.
C source files are generated from this as part of the build process.
First an executable called pgen is built, then pgen is run to
regenerate graminit.*.
In a cross-compile situation this goes terribly wrong, pgen is
compiled to the host architecture, meaning it's unable to run on
the build architecture.
Luckily the tarball ships pregenerated graminit.*, and we haven't modified
the Grammar file. So we can just copy across and force make to use them.

moreati committed:
Python: remove duplicated unsafe_string_sources entries

moreati committed:
Python: Add another strcpy() to unsafe_string_sources

moreati committed:
Python: Patch two more references to open()

moreati committed:
Python: prevent autoconf trying to fill in st_blocks

moreati committed:
Python: Make the build deterministic

moreati committed:
Python: keep PyErr_CheckSignals as a nop

This satsifies the linker, without having to patch calls throughout the
codebase.

Please excuse the excessive line noise in this diff. I'm still at the get
it working phase.

moreati committed:
Python: patch out another getpid() reference

moreati committed:
Python: prevent the pwd module getting built

moreati committed:
Python: Copy importlib.h et al from dist tarball

This is to prevent the Makefile trying (and failing) to regenerate it.

moreati committed:
Python: patch configure.ac to match configure

configure.ac is not actually invoked by the CloudABI BUILD, but
having the changes applied to both prevents potential confusion.

moreati committed:
Python: specify an existing python on the build system

When cross compiling on a BUILD architecture, for another HOST architecture
(e.g. BUILD=x86_64-linux-gnu for HOST=aarch64-unknown-cloudbai) Python's
build scripts require an existing copy of python on the BUILD architecture.
They use this to generate some files (e.g. pybuilddir.txt).

Hardcoding in BUILD is bad, figure out something better later.

moreati committed:
Python: Build a native binary, to support the cross-build

See https://mail.python.org/pipermail/python-dev/2016-February/143368.html

moreati committed:
Python: Add various ./configure checks, use them

e.g. for signal()/HAVE_SIGNAL, replace uses of #if 0

This is necessary because we're now doing a native build (in which we have
full POSIX) and a CloudABI build. From the same source tree.

moreati committed:
Python: Refactor stdin/stdout & printf patches

These patches now use #ifdef to protect function calls,symbols etc that
CloudABI omits. This is so we can build a native binary from the same tree,
as part of the cross-build process.

Note in particular, that this fixes a error we previously introduced in
Parser/accel.c

moreati committed:
Python: whitespace cleanup in patches

moreati committed:
Python cleanup unused import

moreati committed:
Python: Fix PYTHON_FOR_BUILD string conversion

moreati committed:
Python: Declare more unsafe_string_sources

moreati committed:
Python: Do all #define HAVE_* removals together

moreati committed:
Python: More autoconf header/function checks

moreati committed:
Python: Autoconf for termios, resource & nis modules

moreati committed:
Python: Autoconf test to build ossaudiodev module

moreati committed:
Python: #ifdef various kqueue EVFILT* constants

moreati committed:
Python: #ifdef POLLPRI constant

moreati committed:
Python: Autconf test for grp.h

moreati committed:
Python: #ifdef for AF_UNIX sockaddr_un.sun_path field

moreati committed:
Python: Autoconf tests for flock(), F_* & LOCK_*

moreati committed:
Python: Another umask() reference

moreati committed:
Python: Another reference to access()

moreati committed:
Python: Another reference to chdir()

moreati committed:
Python: Another reference to chmod()

moreati committed:
Python: Autoconf test for clock_settime()

moreati committed:
Python: More references to fopen()

moreati committed:
Python: Another reference to fork()

moreati committed:
Python: Another reference to getcwd()

moreati committed:
Python: gmtime() localtime() refs in datetime module

moreati committed:
Python: Another reference to opendir()

moreati committed:
Python: More references to printf()

moreati committed:
Python: Include pwd in missing modules

moreati committed:
Python: Include autoconf defines in fileutils.h

moreati committed:
Python: Another reference to getpid()

moreati committed:
Python: Use AUTOCONF tests for setlocale()

moreati committed:
Python: More references to getuid()

moreati committed:
Python: More references to signal() et al

moreati committed:
Python: References to execv() and execve()

moreati committed:
Python: Don't assume sem_open() therefore semaphore.h

moreati committed:
Python: CloudABI has no LOG_UPTO in syslog.h

moreati committed:
Python: Fix errant HAVE_SP* checks in spwd module

moreati committed:
Python: CloudABI has no netinet/tcp.h

moreati committed:
Python: Another reference to stdin

moreati committed:
Python: Tighten up HAVE_STAT and HAVE_LSTAT checks

moreati committed:
Python: Don't, #ifdef(somefunction), it's never true

moreati committed:
Python: whitespace cleanup

moreati committed:
Python: Override PYTHON_FOR_BUILD correctly

If the environment variables $_PYTHON_PROJECT_BASE, $_PYTHON_HOST_PLATFORM, &
$PYTHONPATH are not set then when python's build script emits native values
for various config.h values and parameters. Needless to say we want the cross
build values.

The value we're using is the value set in configure.ac, after a search of $PATH
for a Python executable. Within cloudabi-ports that loop never succeeds,
because we don't include /usr/bin et al in `$PATH`

moreati committed:
Python: Disable any use of semaphores

moreati committed:
Python: build statically

This is probably incomplete, but it's a start

moreati committed:
Python: Disable various socket.h & netdb.h bits

This includes:
	SOCK_RAW
	bind()
	connect()
	freehostent() and anything else using struct hostent
	gethost*()
	inet_addr()
	inet_aton()
	inet_ntoa()
	sendto()
	setsockopt()

moreati committed:
Python: Detect more functions in ./configure

moreati committed:
Python: disable crypt module, CloudABI has no crypt()

moreati committed:
Python: Disable srand() seeding in expat module

CloudABI rand() always returns a seeded value

moreati committed:
Python: More printf() and stdin/stdout references

moreati committed:
Python: More fopen() references

moreati committed:
Python: Remove some debugging I accidentally committed

moreati committed:
Python: Don't build/install ensurepip

Purely in the interests of simplicity, and getting something out the door.

moreati committed:
Discard patches no longer needed

moreati committed:
Switch to Python 3.6 dev

moreati committed:
More unsafe strings

moreati committed:
Resynchronise patches with Python 3.6 dev

moreati committed:
Patch out references to rlimit() et al

moreati committed:
Build from Python 3.6.0 alpha2, released yesterday

moreati committed:
Fix possible int overflow

moreati committed:
Fix possibly wrong use #ifdef stdin,  #ifdef stabdout

moreati committed:
Make sure stdin_is_interactive gets initialised

moreati committed:
Build with pydebug (for now)

In the interest of getting *something* that works, and aiding the early porting work.

Not intended for the long term.

moreati committed:
Build without threads (for now)

Again this isn't intended long term, just to simplify the early porting.

Note to whoever removes this, patch-pthread_key_t probably needs attention.
Particularly if it's still doing

+#if !defined(__CloudABI__)
+typedef int PyThread_type_key;
+#else
+typedef intptr_t PyThread_type_key;
+#endif

moreati committed:
Replace all(?) debug calls to printf(...) with fprintf(stderr ...)

moreati committed:
Don't add -static-libgcc to LDFLAGS

We're using clang, and it doesn't need it or want it.

moreati committed:
More debug printf(...) -> fprintf(stderr, ...) replacements

moreati committed:
Set sys.executable, sys.prefix et al to None

This may or may not be the final behaviour, for now it resolves segfaults
that occur when joinpath() is called to construct these values.

moreati committed:
Provide a working stderr for use during porting

This will _not_ be the final behaviour, but it helps with debugging the initial
port. Inspired by packages/lua/patch-tool-lua

moreati committed:
Remove LOG_UPTO patch, cloudlibc now includes it

moreati committed:
Catch attempt to use os.getcwd() in importlib

This avoids the error

  AttributeError: module 'posix' has no attribute 'getcwd',

encountered whilst trying to initialise the default FS encoding.
Allowing us to reach the meat of the problem,

  ImportError: No module named 'encodings'

moreati committed:
Hardcode an empty sys.path

Not final, just enough to reach the next error in the startup sequence.

moreati committed:
Don't use sizeof(addrinfo.sa_data), it's undefined when the type is char[]

moreati committed:
Update BUILD for renamed autoconf()->gnu_configure() methods

moreati committed:
Remove obsolete cross compiling workarounds

As of 3.6.0a2 Python's Makefile doesn't need  these workarounds

moreati committed:
Generate importlib.h out of tree & ensure it is referenced there

Python requires two files (importlib.h & importlib_external.h) which are both
generated (from _bootstrap.py & _bootstrap_external.h). However, during an
out of tree build the regenerated copy is not referenced.

To fix this we add -IPython to the CPP flags. To make sure our regenerated
files are used we delete the versions shipped in the tarball.

moreati committed:
Add rudimentary program_main() function

Finally it's starting to take shape. We can execute

cloudabi-run path/to/python << END
cloudabi-run _obj/install/x86_64-unknown-cloudabi/python/bin/python3.6 << EOF
%TAG ! tag:nuxi.nl,2015:cloudabi/
---
stderr: !fd stderr
path: !file
  path: /home/alex/src/cloudabi-ports/_obj/install/x86_64-unknown-cloudabi/python/lib/python3.6/
command: pass
EOF

and get the result

path: -27095040
/usr/obj/cloudabi-ports/build/0/Python-3.6.0a2/Python/pylifecycle.c:191 _PyCodec_Lookup
/usr/obj/cloudabi-ports/build/0/Python-3.6.0a2/Python/codecs.c:114
/usr/obj/cloudabi-ports/build/0/Python-3.6.0a2/Python/codecs.c:118
/usr/obj/cloudabi-ports/build/0/Python-3.6.0a2/Python/codecs.c:1532
/usr/obj/cloudabi-ports/build/0/Python-3.6.0a2/Python/codecs.c:1557
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ImportError: No module named 'encodings'

Current thread 0x0000000000000000 (most recent call first):
Aborted

moreati committed:
grammar

moreati committed:
Restore os.access(), since it can take a dir_fd parameter

moreati committed:
Inlcude os.link(), upstream only checks for HAVE_LINK

moreati committed:
Set errno when non-existant functions would have been called

Hopefully this will lead to tracebacks that stick out like a sore thumb.

moreati committed:
Exclude the setlocale.setlocale() docstring

moreati committed:
tag #endif lines with the test they match

moreati committed:
Simplify/coalesce the stdin & stdout #ifdef blocks

moreati committed:
WIP Enable import of .py files relative to a file descriptor

To achieve that this changeset

 - Accepts a single file descriptor to a dir, through program_main() & argdata.h
 - Stores that fd as the only item in sys.path
 - Mutilates importlib._bootstrap_external to perform fd-relative imports

moreati committed:
Handle command argdata and other flags

NB: this only checks for the presence of the flags, the values are ignored

moreati committed:
Cleanup patches

moreati committed:
Don't patch out os.chmod(), it can take a dir_fd

moreati committed:
Remove debug_shell() invocation

moreati committed:
Query the Python running build_packages.py for the build arch

moreati committed:
Merge patch-no-setlocale2 into patch-no-setlocale

moreati committed:
Moving overlapping Py_Main() patching to patch-tool-python

moreati committed:
Run fixup_patches.py, move clashing HAVE_* patches into patch-configure

moreati committed:
Use correct sysconfig variable for --build argument in cross build

EdSchouten commented on packages/python/BUILD:
```diff
@@ -0,0 +1,89 @@
+
+    # Delete the in-tree copy of these, to make certain the out-of-tree
+    # versions are used, i.e. those generated from our patched importlib
+    os.unlink(os.path.join(str(srcdir), 'Python', 'importlib.h'))
```
You should be able to write this as just:

```
srcdir.path('Python/importlib.h').remove()
srcdir.path('Python/importlib_external.h').remove()
```

EdSchouten commented on packages/python/BUILD:
```diff
@@ -0,0 +1,89 @@
+    hostbuild.make()
+
+    build = srcdir.gnu_configure([
+        '--build=%s' % sysconfig.get_config_var('HOST_GNU_TYPE'),
```
I'm a bit in doubt what to do here. Some time ago, I added `--build` to the configure flags in `src/builder.py`, but I noticed that this then trickles into the package contents (e.g., part of `bin/...-config`). Maybe the right thing to do would be to not pass this in as a flag, but add a parameter to the function: `gnu_configure(add_build=True)`. I'd rather not solve all of this in this package specifically.

What are your thoughts?

moreati replied:
This line is present because CPython requires `--build=...` to complete a cross-build. On the basis of "explicit it better than implicit' I prefer explicitly building the string to a keyword argument. It might be an idea for the `ctx` object to have e.g. `host_arch`, `build_arch` properties or methods. Rather than import `sysconfig`.

EdSchouten replied:
What about ctx.host().gnu_triple()? I can add that if you want.

moreati replied:
That sounds good

EdSchouten replied:
Done! \o/

EdSchouten commented on packages/python/BUILD:
```diff
@@ -0,0 +1,89 @@
+                                            '&& echo $(abs_builddir)/`cat pybuilddir.txt`:)'
+                                    '$(srcdir)/Lib:'
+                                    '$(srcdir)/Lib/$(PLATDIR) '
+                         '%s' % os.path.join(str(hostbuild), 'python'),
```
`hostbuild.path('python')`

EdSchouten commented on packages/python/patch-addrinfo_sa_data:
```diff
@@ -0,0 +1,19 @@
+                              sizeof(addr->sa_data));
++#else
++        /* CPython assumes addrinfo.sa_data is declared as char[n], for some
++           finite n, but cloudlibc declares addrinfo.sa_data as char[] */
```
That's interesting that Python makes use of this. I'm also quite torn what to do here. Should we declare our `sa_data[]` to also have some random size? Maybe Python should be patched up to not use it at all, even on other platforms?

moreati replied:
I haven't investigated the design motivations on either side, so I have no opinion either way

EdSchouten commented on packages/python/patch-configure:
```diff
@@ -0,0 +1,714 @@
+  wcscoll wcsftime wcsxfrm wmemcmp writev _getpty)
+ 
+ AC_CHECK_DECL(dirfd,
+--- Modules/posixmodule.c
```
Are these changes supposed to be part of this patch file? With `patch-configure` I'd argue that we should only use it for changes to build infrastructure.

moreati replied:
Do you mean the below removals of `#define HAVE_*` and `#if defined(HAVE_*)`? If so then they're here because they're in such close proximity and the patch hunks were clashing

EdSchouten replied:
Ah, got it. Makes sense then.

EdSchouten commented on packages/python/patch-no-access:
```diff
@@ -0,0 +1,14 @@
++#ifdef HAVE_ACCESS
+         result = access(path->narrow, mode);
++#else
++        {result = -1; errno = EFAULT;}
```
Could you maybe let this use `ENOSYS` instead of `EFAULT`? `EFAULT` may be misleading, because there is no actual memory fault occurring. There are also other errnos that we return elsewhere that should all be `ENOSYS`.

Even better would be to return `ENOTCAPABLE`, but that's CloudABI/FreeBSD+Capsicum specific.

moreati replied:
I chose the error code semi-randomly, to be 'unusual` so it would be easier to cross correlate exceptions with lines of code. I'll change them to be ENOSYS.

EdSchouten commented on packages/python/patch-no-opendir:
```diff
@@ -0,0 +1,36 @@
+ 
+         Py_BEGIN_ALLOW_THREADS
++#ifdef HAVE_OPENDIR
+         dirp = opendir(name);
```
Is this safe to do? As in, is `dirp` initialised with `NULL` elsewhere? Maybe the entire function should be disabled? Same for the other changes in this file.

moreati replied:
I'll investigate

moreati replied:
It is safe, `dirp` is initialised to `NULL`, and next line performs an `if (dirp == NULL)` check

EdSchouten commented on packages/python/patch-no-srand:
```diff
@@ -0,0 +1,23 @@
+ #include <string.h>                     /* memset(), memcpy() */
+ #include <assert.h>
+ #include <limits.h>                     /* UINT_MAX */
++#if !defined(__CloudABI__)
```
This chunk of the diff can be removed, I guess. There is no harm with including `<time.h>`.

We also had a similar patch in `packages/expat`, but the latest version of expat will require a different patch anyway. They changed the PRNG to use something equally bad.

moreati replied:
I'm not sure I understand. Why include `time.h` when I've `#ifdef` disable the only use of it?

EdSchouten replied:
It has the advantage that it reduces the total size of the diff, meaning it's less likely to break when updating (e.g., when upstream adds additional includes).

EdSchouten commented on packages/python/patch-no-sys-resource_h:
```diff
@@ -0,0 +1,29 @@
+--- Modules/_lsprof.c
```
It may be possible to remove this patch nowadays. We recently gained a copy of `<sys/resource.h>` that includes a subset of the features of POSIX.

moreati replied:
I've created https://github.com/moreati/cloudabi-ports/issues/4 to record this. Would you accept the PR with this as is, for now?

EdSchouten replied:
Sure!

EdSchouten commented on packages/python/patch-pthread_key_t:
```diff
@@ -0,0 +1,269 @@
+     if (fail)
+         return -1;
++#if !defined(__CloudABI__)
+     if (key > INT_MAX) {
```
Now that `PyThread_type_key` has been changed to an `intmax_t`, this check can now be removed, properly fixing issue 22206. 👍

moreati replied:
The `PyThread_type_key` stuff is almost certainly broken. I added it back in March when I was trying to get Python to even compile. IIRC CPython relies on the fact that pthread handles happen to be of type `int` on most POSIX  operating systems. It's probably better to remove this patch entirely. The BUILD file already disables threading

moreati replied:
I've removed the pthread_type_key patch and created https://github.com/moreati/cloudabi-ports/issues/5 to record the lack of threading support

EdSchouten commented on packages/python/patch-pthread_key_t:
```diff
@@ -0,0 +1,269 @@
+-    return (int)key;
++    return (PyThread_type_key)key;
++#else
++    return (PyThread_type_key)&key;
```
This won't work. We're returning an object that has been allocated on the stack by reference. We should first allocate the `pthread_key_t` using `malloc()` and call `pthread_key_create()` on that.

EdSchouten commented on packages/python/patch-tm_zone:
```diff
@@ -0,0 +1,14 @@
+--- Modules/timemodule.c
```
It's quite annoying that this is needed. Cloudlibc declares the non-standard `tm_zone` different from other C libraries. In our case it's an array instead of a pointer. The reason for this is that time zone strings are generated on the spot by `localtime_l()`. If we want it to be a pointer, we need to change it to use a string pool internally. Blegh.

EdSchouten commented on packages/python/BUILD:
```diff
@@ -0,0 +1,89 @@
+        'ac_cv_file__dev_ptc=no',
+        'ac_cv_file__dev_ptmx=no',
+        'ac_cv_have_long_long_format=yes',
+        'ac_cv_posix_semaphores_enabled=no',
```
Is this disabled, because Python depends on the named POSIX semaphores (`sem_open`)? We do provide support for the anonymous ones (`sem_init`).

moreati replied:
It's one of the earliest lines that I added. IIRC I cargo culted it from another Python cross compilation script. I'll try without it and see what happens.

moreati replied:
Removing `ac_cv_posix_semaphores_enabled=no` didn't break Hello World, so I've done so.

moreati committed:
Use FileHandle.path() to reference source & build artifacts

moreati committed:
Sett errno = ENOSYS when Python code tries to call removed libc functions

See https://github.com/NuxiNL/cloudabi-ports/pull/14#discussion_r70165101

moreati committed:
Disable broken thread handle handling

moreati committed:
Don't assume ac_cv_posix_semaphores_enabled=no

I inlcuded this line in the very early days of this port, from another cross-build recipe, without knowing exactly why it was needed. Since the build works without it I'm removing it.

moreati committed:
Merge remote-tracking branch 'upstream/master' into python

moreati committed:
Query host system using BuildHandle methods, remove sysconfig import

moreati committed:
Don't patch out time.h include

It's not strictly necessary, and it makes the patch more rebust to changes in upstream

moreati commented:
I think that covers all the comments

EdSchouten commented on packages/python/BUILD:
```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2015 Nuxi, https://nuxi.nl/
```
You forgot to add your own copyright to this file. ;-)

EdSchouten commented on packages/python/BUILD:
```diff
@@ -0,0 +1,86 @@
+    name='python',
+    version='3.6.0',
+    homepage='https://www.python.org/',
+    maintainer='info@nuxi.nl',
```
If you want, you can put your own email address here. :-)

EdSchouten commented:
Splendid! I think the package is _almost_ ready for integration. Be sure to address those two final comments and I'll merge this.

moreati committed:
Add copyright and set myself as package maintainer

moreati commented:
On 10 July 2016 at 08:17, EdSchouten notifications@github.com wrote:

> Splendid! I think the package is _almost_ ready for integration. Be sure
> to address those two final comments and I'll merge this.
> 
> Done, thanks for all the review work.

moreati commented:
Thank you :smile:
